lib/galaxy/auth/providers/ldap_ad.py

```diff
@@ -282,11 +295,179 @@ def authenticate_user(self, user, password, options, request):
+                conn.search(
+                    search_base=_get_subs(options, "search-base", params),
+                    search_scope=ldap3.SUBTREE,
+                    search_filter=_get_subs(options, "search-filter", params),
+                    attributes=attributes,
+                    time_limit=60,
+                    size_limit=1,
+                )
+                response = conn.response
+                # Unbind connection
+                conn.unbind()
+
+                # parse results
+                if len(response) == 0 or "attributes" not in response[0].keys():
+                    log.warning("LDAP3 authenticate: search returned no results")
+                    return (failure_mode, None)
+                dn = response[0]["dn"]
+                attrs = response[0]["attributes"]
+                log.debug("LDAP3 authenticate: dn is %s", dn)
+                log.debug("LDAP3 authenticate: search attributes are %s", attrs)
+                for attr in attributes:
+                    if self.role_search_attribute and attr == self.role_search_attribute[1:-1]:  # strip curly brackets
+                        # keep role names as list
+                        params[self.role_search_option] = [unicodify(_) for _ in attrs[attr]]
+                    elif attr == "memberOf":
+                        params[attr] = [unicodify(_) for _ in attrs[attr]]
+                    elif attr in attrs:
+                        params[attr] = unicodify(attrs[attr][0])
+                    else:
+                        params[attr] = ""
+
+                if self.auto_create_roles_or_groups and self.role_search_option not in params:
+                    raise ConfigurationError(
+                        "Missing or mismatching LDAP parameters for %s. Make sure the %s is "
+                        "included in the 'search-fields'." % (self.role_search_option, self.role_search_attribute)
+                    )
+                params["dn"] = dn
+            except Exception:
+                log.exception("LDAP3 authenticate: search exception")
+                return (failure_mode, None)
+
+        return failure_mode, params
+
+    def _authenticate(self, params, options):
+        """
+        Do the actual authentication by binding as the user to check their credentials
+        """
+        try:
+            # Initialise server object
+            self.get_server(options, params)
+            conn = ldap3.Connection(
+                self.server,
+                user=_get_subs(options, "bind-user", params),
+                password=_get_subs(options, "bind-password", params),
+                auto_bind=self.auto_bind,
+            )
+            # Use StartTLS if LDAP connection is not over TLS
+            if not self.ldap_tls:
+                conn.start_tls()
+            try:
+                whoami = conn.extend.standard.who_am_i()
+                # Unbind connection
+                conn.unbind()
+            except Exception:
```

Comment:
Can you restrict this exception?

Reply:
Cheers. Nice find.  We need to catch [`LDAPExtensionError`](https://github.com/cannatag/ldap3/blob/78c449dded29dad54e4e68055f7290ad8689fbd1/test/testExtendedOperations.py#L57-L65). Will change it.